src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -179,18 +191,40 @@ private void EnqueueParseSnapshotTask(Document newDocument)
                 // we will enqueue new one soon, cancel pending refresh right away
                 _reportChangeCancellationSource.Cancel();
 
-                var newText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
-                var snapshot = newText.FindCorrespondingEditorTextSnapshot();
-                if (snapshot == null)
+                ITextSnapshot snapshot;
+                if (document != null)
                 {
-                    // It's possible that we're seeing a notification for an update that happened
-                    // just before the file was opened, and so the document we're given is still the
-                    // old one.
-                    return;
+                    var newText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                    snapshot = newText.FindCorrespondingEditorTextSnapshot();
+                    if (snapshot == null)
+                    {
+                        // It's possible that we're seeing a notification for an update that happened
+                        // just before the file was opened, and so the document we're given is still the
+                        // old one.
+                        return;
+                    }
+
+                    // preemptively parse file in background so that when we are called from tagger from UI thread, we have tree ready.
+                    var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
See `[GetSyntaxTreeAsync](http://source.roslyn.io/Microsoft.CodeAnalysis.Workspaces/a.html#5c64b7a58e6a3d72)`.  The document caches the result.